storage/storage_transport.go

```diff
@@ -0,0 +1,86 @@
+package storage
+
+import (
+	"errors"
+	"regexp"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/containers/image/docker/reference"
+	"github.com/containers/image/types"
+	"github.com/containers/storage/storage"
+)
+
+var (
+	// Transport is an ImageTransport that uses a default storage.Store or
+	// one that's it's explicitly told to use.
+	Transport StoreTransport = &storageTransport{}
+	// ErrInvalidReference is returned when ParseReference() is passed an
+	// empty reference.
+	ErrInvalidReference = errors.New("invalid reference")
+	idRegexp            = regexp.MustCompile("^(sha256:)?([0-9a-fA-F]{64})$")
+)
+
+// StoreTransport is an ImageTransport that uses a storage.Store to parse
+// references, either its own or one that it's told to use.
+type StoreTransport interface {
+	types.ImageTransport
+	SetStore(storage.Store)
+	ParseStoreReference(store storage.Store, reference string) (types.ImageReference, error)
+}
+
+type storageTransport struct {
+	store storage.Store
+}
+
+func (s *storageTransport) Name() string {
+	return "storage"
+}
+
+// SetStore sets the Store object which the Transport will use for parsing
+// references when a Store is not directly specified.  If one is not set, the
+// library will attempt to initialize one with default settings when a
+// reference needs to be parsed.  Calling SetStore does not affect previously
+// parsed references.
+func (s *storageTransport) SetStore(store storage.Store) {
+	s.store = store
+}
+
+// ParseStoreReference takes a name or an ID, tries to figure out which it is
+// relative to a given store, and returns it in a reference object.
+func (s *storageTransport) ParseStoreReference(store storage.Store, ref string) (types.ImageReference, error) {
+	if ref == "" {
+		return nil, ErrInvalidReference
+	}
+	digest, name, err := reference.ParseIDOrReference(ref)
+	if err != nil {
+		return nil, err
+	}
+	id := ""
+	if name == nil {
+		ref = ""
+		id = digest.Hex()
+		logrus.Debugf("parsed reference %q into ID %q", ref, digest)
+	} else {
+		id = ""
+		ref = reference.WithDefaultTag(name).String()
+		logrus.Debugf("treating reference %q as a name", ref)
+	}
+	return newReference(store, s, ref, id), nil
+}
+
+// ParseReference initializes the storage library and then takes a name or an
+// ID, tries to figure out which it is, and returns it in a reference object.
+func (s *storageTransport) ParseReference(reference string) (types.ImageReference, error) {
+	if s.store == nil {
+		store, err := storage.MakeStore(storage.DefaultStoreOptions)
+		if err != nil {
+			return nil, err
+		}
+		s.store = store
+	}
+	return s.ParseStoreReference(s.store, reference)
+}
+
+func (s *storageTransport) ValidatePolicyConfigurationScope(scope string) error {
+	return nil
```

Comment:
Okay, if references can include store location prefixes, then a store prefix by itself is a valid namespace name if we want to be able to specify a policy that covers every image in that store, so I'm fleshing this one out along those lines.